src/Symfony/Component/Mailer/Bridge/Mailchimp/RemoteEvent/MailchimpPayloadConverter.php

```diff
@@ -0,0 +1,73 @@
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Mailer\Bridge\Mailchimp\RemoteEvent;
+
+use Symfony\Component\RemoteEvent\Event\Mailer\AbstractMailerEvent;
+use Symfony\Component\RemoteEvent\Event\Mailer\MailerDeliveryEvent;
+use Symfony\Component\RemoteEvent\Event\Mailer\MailerEngagementEvent;
+use Symfony\Component\RemoteEvent\Exception\ParseException;
+use Symfony\Component\RemoteEvent\PayloadConverterInterface;
+
+final class MailchimpPayloadConverter implements PayloadConverterInterface
+{
+    public function convert(array $payload): AbstractMailerEvent
+    {
+        if (!\in_array($payload['event'], ['send', 'deferral', 'hard_bounce', 'soft_bounce', 'delivered', 'open', 'click', 'spam', 'unsub', 'reject'], true)) {
+            throw new ParseException(sprintf('Unsupported event "%s".', $payload['event']));
+        }
+        if (\in_array($payload['event'], ['send', 'deferral', 'soft_bounce', 'hard_bounce', 'delivered', 'reject'], true)) {
+            $name = match ($payload['event']) {
+                'send' => MailerDeliveryEvent::RECEIVED,
+                'deferral', 'soft_bounce' => MailerDeliveryEvent::DEFERRED,
+                'hard_bounce' => MailerDeliveryEvent::BOUNCE,
+                'delivered' => MailerDeliveryEvent::DELIVERED,
+                'reject' => MailerDeliveryEvent::DROPPED,
+            };
+
+            $event = new MailerDeliveryEvent($name, $payload['msg']['_id'], $payload);
+            // reason is only available on failed messages
+            $event->setReason($this->getReason($payload));
+        } else {
+            $name = match ($payload['event']) {
+                'click' => MailerEngagementEvent::CLICK,
+                'open' => MailerEngagementEvent::OPEN,
+                'spam' => MailerEngagementEvent::SPAM,
+                'unsub' => MailerEngagementEvent::UNSUBSCRIBE,
+            };
+            $event = new MailerEngagementEvent($name, $payload['msg']['_id'], $payload);
+        }
+
+        if (!$date = \DateTimeImmutable::createFromFormat('U', $payload['msg']['ts'])) {
+            throw new ParseException(sprintf('Invalid date "%s".', $payload['msg']['ts']));
+        }
+        $event->setDate($date);
+        $event->setRecipientEmail($payload['msg']['email']);
+        $event->setMetadata($payload['msg']['metadata']);
+        $event->setTags($payload['msg']['tags']);
+
+        return $event;
+    }
+
+    private function getReason(array $payload): string
+    {
+        if (null !== $payload['msg']['diag']) {
+            return $payload['msg']['diag'];
+        }
+        if (null !== $payload['msg']['bounce_description']) {
+            return $payload['msg']['bounce_description'];
+        }
+        if (null !== $payload['msg']['smtp_events'] && null !== $payload['msg']['smtp_events']['diag']) {
+            return $payload['msg']['smtp_events']['diag'];
+        }
```

Comment:
According to https://mailchimp.com/developer/transactional/docs/webhooks/#detailed-webhook-format-responses, is an array of object right ?

Reply:
Your right `an array of JSON objects, each of which is an SMTP response received for the message. Each item in the array will contain the following keys: `

I will have to update to read every `diag` and concat them together